plugins/user_mapping/src/local_file.cpp

```diff
@@ -0,0 +1,122 @@
+#include "interface.h"
+#include "local_file.hpp"
+
+#include <cstdlib>
+#include <cstring>
+#include <exception>
+#include <fstream>
+#include <mutex>
+#include <shared_mutex>
+#include <string>
+#include <vector>
+
+#include <spdlog/spdlog.h>
+
+namespace
+{
+	std::string file_path;
+	std::vector<UserProfile> profile_list;
+	std::shared_mutex list_mutex;
+} // anonymous namespace
+
+auto init(const nlohmann::json& _config) -> void {
+	const auto path{_config.find("file_path")};
+	if (path == std::end(_config)) {
+		// handle bad case
+		spdlog::error("{}: Unable to find 'file_path' in provided config.", __func__);
+	}
+
+	// Save file string
+	file_path = *path;
+
+	// Parse into json object
+	spdlog::trace("{}: openning file_path [{}].", __func__, file_path);
+	update();
+}
+
+auto update() -> void {
+	try {
+		std::unique_lock update_lock{list_mutex, std::defer_lock};
+
+		std::ifstream file{file_path};
+		spdlog::debug("{}: Opened file [{}]", __func__, file_path);
+
+		auto json_data{nlohmann::json::parse(file)};
+
+		// Create temp list to be swapped in later
+		std::vector<UserProfile> temp_list;
+		temp_list.reserve(json_data.size());
+
+		// Process into list:
+		auto base_itter{json_data.items()};
+		std::transform(std::begin(base_itter), std::end(base_itter), std::back_inserter(temp_list), [](const auto& _itter) -> UserProfile {
+			return {.irods_user_name = _itter.key(), .attributes = _itter.value()};
+		});
+
+		// Move new item list to old one
+		if (update_lock.try_lock()) {
+			profile_list = std::move(temp_list);
+		}
+	} catch(const std::exception& e) {
+		spdlog::debug("{}: {}", __func__, e.what());
+	}
```

Comment:
Is this for catching the exception thrown by dereferencing the `std::end` iterator in `init` (L18-20)? Are there any other exceptions which could be thrown that we don't want to mask?

Reply:
I see. And this function returns `void` so I guess there's no opportunity to communicate "real" errors... If we don't expect this to cause any problems then I guess this is fine.

Reply:
We could always allow for the returning of a status code in the functions. Then, at the call site, check for a non-zero code, meaning there is an error.

Reply:
Will implement error code return for the functions, move c-string returns to out variable.

Reply:
Error codes implemented. try/catch in impl functions removed to allow 'interface functions' to handle error codes.